monitoring/nethealth.go

```diff
@@ -0,0 +1,411 @@
+	"github.com/gravitational/satellite/agent/health"
+
+	"github.com/gravitational/roundtrip"
+	"github.com/gravitational/trace"
+	"github.com/mailgun/holster"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NethealthConfig specifies configuration for a nethealth checker.
+type NethealthConfig struct {
+	// AdvertiseIP specifies the advertised ip address of the host running this checker.
+	AdvertiseIP string
+	// NethealthPort specifies the port that nethealth is listening on.
+	NethealthPort int
+	// NetStatsInterval specifies the duration to store net stats.
+	NetStatsInterval time.Duration
+	// KubeConfig specifies kubernetes access information.
+	*KubeConfig
+}
+
+// CheckAndSetDefaults validates that this configuration is correct and sets
+// value defaults where necessary.
+func (c *NethealthConfig) CheckAndSetDefaults() error {
+	var errors []error
+	if c.AdvertiseIP == "" {
+		errors = append(errors, trace.BadParameter("host advertise ip must be provided"))
+	}
+	if c.NethealthPort == 0 {
+		c.NethealthPort = defaultNethealthPort
+	}
+	if c.NetStatsInterval == time.Duration(0) {
+		c.NetStatsInterval = defaultNetStatsInterval
+	}
+	if c.KubeConfig == nil {
+		errors = append(errors, trace.BadParameter("kubernetes access config must be provided"))
+	}
+	return trace.NewAggregate(errors...)
+}
+
+// nethealthChecker checks network communication between peers.
+type nethealthChecker struct {
+	// NethealthConfig contains caller specified nethealth checker configuration
+	// values.
+	NethealthConfig
+	// seriesCapacity specifies the number of data points to store in a series.
+	seriesCapacity int
+	// lock access to timeoutStats
+	sync.Mutex
+	// peerStats maps a peer to its recorded nethealth stats.
+	peerStats netStats
+}
+
+// NewNethealthChecker returns a new nethealth checker.
+func NewNethealthChecker(config NethealthConfig) (*nethealthChecker, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// seriesCapacity is determined by the NetStatsInterval / StatusUpdateTimeout
+	// rounded up to the nearest integer
+	seriesCapacity := math.Ceil(float64(config.NetStatsInterval) / float64(agent.StatusUpdateTimeout))
+
+	return &nethealthChecker{
+		peerStats:       newNetStats(netStatsCapacity),
+		seriesCapacity:  int(seriesCapacity),
+		NethealthConfig: config,
+	}, nil
+}
+
+// Name returns this checker name
+// Implements health.Checker
+func (c *nethealthChecker) Name() string {
+	return nethealthCheckerID
+}
+
+// Check verifies the network is healthy.
+// Implements health.Checker
+func (c *nethealthChecker) Check(ctx context.Context, reporter health.Reporter) {
+	probes, err := c.check(ctx)
+	if err != nil {
+		log.WithError(err).Error("Unable to verify nethealth.")
+	}
+
+	if probes.NumProbes() != 0 {
```

Comment:
To be honest, instead of returning probes here, I would still pass reporter to check() function like it was before, but just made sure that check() returns "nil" in case nethealth pod wasn't found. This way it would plug seamlessly into the old logic and it would be cleaner because probes would only be created in a single place. For example:

```go
func (c *nethealthChecker) Check(ctx context.Context, reporter health.Reporter) {
  probes, err := c.check(ctx)
  if err != nil {
    // add error probe
  }
  // add success probe
}

func (c *nethealthChecker) check(ctx context.Context, reporter health.Reporter) error {
  addr, err := c.getNethealthAddr(pods)
  if err != nil && !trace.IsNotFound(err) {
    return trace.Wrap(err) // we got unexpected error, maybe networking-related, will add error probe above
  }
  if trace.IsNotFound(err) {
    log.Debug("Nethealth pod not found.")
    return nil // pod wasn't found, log and treat gracefully
  }
  //...
}

func (c *nethealthChecker) getNethealthAddr() (string, error) {
  pods, err := c.Client.CoreV1().Pods(nethealthNamespace).List(metav1.ListOptions{})
  if err != nil {
    return "", rigging.ConvertError(err) // this will convert error to a proper trace error, e.g. trace.NotFound
  }
  // ...
}
```

Reply:
Ah ok, reverting back to previous structure. 
Currently unable to import a compatible version of `rigging`. Opened a PR to bump k8s and set a trace version https://github.com/gravitational/rigging/pull/75

Reply:
Hmm, since we need this in 5.5 I'm not sure we can safely bump Kubernetes version in rigging and revendor here. Might run into incompatibility issues - 5.5 has Kubernetes 1.13.

Reply:
@bernardjkim Btw on that note - I have just noticed that this PR is open to master. Shouldn't it be pointing to version/5.5.x instead? Because we're essentially backporting all this to 5.5 so all code changes should go to version/5.5.x branches of respective codebases (satellite, planet, etc.)